lib/SPIRV/libSPIRV/SPIRVValue.h

```diff
@@ -148,6 +149,19 @@ template <spv::Op OC> class SPIRVConstantBase : public SPIRVValue {
     recalculateWordCount();
     validate();
   }
+  // Complete constructor for AP integer constant
+  SPIRVConstantBase(SPIRVModule *M, SPIRVType *TheType, SPIRVId TheId,
+                    llvm::APInt TheValue)
+      : SPIRVValue(M, 0, OC, TheType, TheId) {
+    const uint64_t *BigValArr = TheValue.getRawData();
+    for (size_t I = 0; I != TheValue.getNumWords(); ++I) {
+      Union.Words[I * 2 + 1] =
+          (uint32_t)((BigValArr[I] & 0xFFFFFFFF00000000LL) >> 32);
+      Union.Words[I * 2] = (uint32_t)(BigValArr[I] & 0xFFFFFFFFLL);
+    }
```

Comment:
What about using `std::memcpy` here?

Reply:
Are you sure we get correct SPIR-V representation for the constant literal? According to https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#OpConstant :

> Value is the bit pattern for the constant. Types 32 bits wide or smaller take one word. Larger types take multiple words, with low-order words appearing first.

For `i96` for example, does `BigValArr[2]` appear first? Am I missing something?

Reply:
Probably `I` should be incremented more, i.e. `I += 2`, as we fill two words within a single iteration. I still think that `std::memcpy` is better. Or, we could do `reinterpret_cast` to align types and then perform a simpler per-element copy using either 32- or 64-bit words to simplify the loop

Reply:
It's possible to implement this via memcpy. When I schemed the change, on paper memcpy was looking more difficult because of the necessity to have the correct SPIRWord's order. I don't mind to try it in the code.

Reply:
> When I schemed the change, on paper memcpy was looking more difficult because of the necessity to have the correct SPIRWord's order.

I've just looked into the spec and indeed lower-order words should appear first. Unfortunately, I can't find any docs about how data is stored within `APInt`. Judging by the code from this PR and from llvm.org it seems that the ordering is the same, but because of the difference in word size we need to swap two halfs of 64-bit `APInt` word before putting it into SPIR-V - right?

If that's the case then I indeed suggest to leave everything as-is

Reply:
Right.